src/printer.js

```diff
@@ -2022,6 +1978,76 @@ function printMemberChain(node, options, print) {
   }
 }
 
+function printJSXElement(path, options, print) {
+  const n = path.getValue();
+  var openingLines = path.call(print, "openingElement");
+
+  let elem;
+  if (n.openingElement.selfClosing) {
+    assert.ok(!n.closingElement);
+
+    elem = openingLines;
+  } else {
+    var children = [];
+
+    path.map(
+      function(childPath) {
+        var child = childPath.getValue();
+
+        if (
+          namedTypes.Literal.check(child) && typeof child.value === "string"
+        ) {
+          if (/\S/.test(child.value)) {
+            const beginBreak = child.value.match(/^\s*\n/);
+            const endBreak = child.value.match(/\n\s*$/);
+
+            children.push(
+              beginBreak ? hardline : "",
+              child.value.replace(/^\s+|\s+$/g, endBreak ? "" : " "),
+              endBreak ? hardline : ""
+            );
+          } else if (/\n/.test(child.value)) {
+            children.push(hardline);
+          }
+        } else {
+          children.push(print(childPath));
+        }
+      },
+      "children"
+    );
+
+    var mostChildren = children.slice(0, -1);
+    var closingLines = path.call(print, "closingElement");
+
+    const firstChild = children[0];
+    const lastChild = util.getLast(children);
+    const beginBreak = firstChild && firstChild.type === "line";
+    const endBreak = lastChild && lastChild.type === "line";
```

Comment:
@rattrayalex Sorry I didn't review this until now.

I'm not entirely sure of the logic you're going for here, but is `util.newlineExistsBefore` helpful? If you're trying to emulate the breaking behavior in the original source, that and `util.newlineExistsAfter` allows you to peek into the original source and see if a node is on a newline or not. Here's how you use it:

```js
util.newlineExistsBefore(text, util.locStart(node))
```

What this does is it scans backward until it hits either a newline or a non-whitespace character. If it hits a newline, it reports `true`, otherwise it's `false`.